backup: TestRestoreAsOfSystemTimeGCBounds failed [potential use of stale sql lease entry]

backup.TestRestoreAsOfSystemTimeGCBounds [failed](https://mesolite.cluster.engflow.com/invocations/default/9fc60dad-cdf3-46cb-a71c-871bb941e9bb?testReportRun=1&testReportShard=5&testReportAttempt=1#targets-Ly9wa2cvYmFja3VwOmJhY2t1cF90ZXN0) on master @ [81f046c5eb268fcf87c8c34fa099dc291cdf24aa](https://github.com/cockroachdb/cockroach/commits/81f046c5eb268fcf87c8c34fa099dc291cdf24aa):


```
=== RUN   TestRestoreAsOfSystemTimeGCBounds
    test_log_scope.go:172: test logs captured to: outputs.zip/logTestRestoreAsOfSystemTimeGCBounds685059863
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:183: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    test_server_shim.go:679: DRPC is enabled (via testserver factory defaults)
    backup_test.go:3895: error executing query="BACKUP TABLE data.bank INTO $1 WITH revision_history" args=["nodelocal://1//tbl-after-gc"]: pq: batch timestamp 1768520688.886393001,0 must be after replica GC threshold 1768520689.053977222,0 (r81: /Tenant/10{-/Table/106/1/1})
    testutils.go:326: no Invalid Descriptors
    panic.go:615: -- test log scope end --
test logs left over in: outputs.zip/logTestRestoreAsOfSystemTimeGCBounds685059863
--- FAIL: TestRestoreAsOfSystemTimeGCBounds (2.77s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=5</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRestoreAsOfSystemTimeGCBounds.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-58750

backup.TestRestoreAsOfSystemTimeGCBounds [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20984512?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20984512?buildTab=artifacts#/) on master @ [81f046c5eb268fcf87c8c34fa099dc291cdf24aa](https://github.com/cockroachdb/cockroach/commits/81f046c5eb268fcf87c8c34fa099dc291cdf24aa):


```
=== RUN   TestRestoreAsOfSystemTimeGCBounds
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/322912c8338911c28b7252db0743d5bf/logTestRestoreAsOfSystemTimeGCBounds4173248955
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:183: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    test_server_shim.go:679: DRPC is enabled (via testserver factory defaults)
    backup_test.go:3895: error executing query="BACKUP TABLE data.bank INTO $1 WITH revision_history" args=["nodelocal://1//tbl-after-gc"]: pq: batch timestamp 1768530713.726530814,0 must be after replica GC threshold 1768530713.929280644,0 (r81: /Tenant/10{-/Table/106/1/1})
    testutils.go:326: no Invalid Descriptors
    panic.go:615: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/322912c8338911c28b7252db0743d5bf/logTestRestoreAsOfSystemTimeGCBounds4173248955
--- FAIL: TestRestoreAsOfSystemTimeGCBounds (1.78s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRestoreAsOfSystemTimeGCBounds.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

backup.TestRestoreAsOfSystemTimeGCBounds [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20985631?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20985631?buildTab=artifacts#/) on master @ [fd8129a1c152c527c43c94d5cb8355ab74e9dd4d](https://github.com/cockroachdb/cockroach/commits/fd8129a1c152c527c43c94d5cb8355ab74e9dd4d):


```
=== RUN   TestRestoreAsOfSystemTimeGCBounds
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/322912c8338911c28b7252db0743d5bf/logTestRestoreAsOfSystemTimeGCBounds3458440312
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:183: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    test_server_shim.go:679: DRPC is enabled (via testserver factory defaults)
    backup_test.go:3895: error executing query="BACKUP TABLE data.bank INTO $1 WITH revision_history" args=["nodelocal://1//tbl-after-gc"]: pq: batch timestamp 1768556656.726443903,0 must be after replica GC threshold 1768556656.834681468,0 (r81: /Tenant/10{-/Table/106/1/1})
    testutils.go:326: no Invalid Descriptors
    panic.go:615: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/322912c8338911c28b7252db0743d5bf/logTestRestoreAsOfSystemTimeGCBounds3458440312
--- FAIL: TestRestoreAsOfSystemTimeGCBounds (1.35s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRestoreAsOfSystemTimeGCBounds.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

backup.TestRestoreAsOfSystemTimeGCBounds [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20986333?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20986333?buildTab=artifacts#/) on master @ [6bf0f9778f34df3d197c28c7f4b2d1a4a27c3288](https://github.com/cockroachdb/cockroach/commits/6bf0f9778f34df3d197c28c7f4b2d1a4a27c3288):


```
=== RUN   TestRestoreAsOfSystemTimeGCBounds
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/322912c8338911c28b7252db0743d5bf/logTestRestoreAsOfSystemTimeGCBounds2871498839
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:183: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    backup_test.go:3895: error executing query="BACKUP TABLE data.bank INTO $1 WITH revision_history" args=["nodelocal://1//tbl-after-gc"]: pq: batch timestamp 1768576232.429121387,0 must be after replica GC threshold 1768576232.534370823,0 (r81: /Tenant/10{-/Table/106/1/1})
    testutils.go:326: no Invalid Descriptors
    panic.go:615: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/322912c8338911c28b7252db0743d5bf/logTestRestoreAsOfSystemTimeGCBounds2871498839
--- FAIL: TestRestoreAsOfSystemTimeGCBounds (3.12s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRestoreAsOfSystemTimeGCBounds.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

backup.TestRestoreAsOfSystemTimeGCBounds [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20986246?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20986246?buildTab=artifacts#/) on master @ [91a24579553850d76c7b538021b3681e39c8af7a](https://github.com/cockroachdb/cockroach/commits/91a24579553850d76c7b538021b3681e39c8af7a):


```
=== RUN   TestRestoreAsOfSystemTimeGCBounds
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/322912c8338911c28b7252db0743d5bf/logTestRestoreAsOfSystemTimeGCBounds3370036045
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:183: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    test_server_shim.go:679: DRPC is enabled (via testserver factory defaults)
    backup_test.go:3895: error executing query="BACKUP TABLE data.bank INTO $1 WITH revision_history" args=["nodelocal://1//tbl-after-gc"]: pq: batch timestamp 1768575873.114573827,0 must be after replica GC threshold 1768575873.470436272,0 (r81: /Tenant/10{-/Table/106/1/1})
    testutils.go:326: no Invalid Descriptors
    panic.go:615: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/322912c8338911c28b7252db0743d5bf/logTestRestoreAsOfSystemTimeGCBounds3370036045
--- FAIL: TestRestoreAsOfSystemTimeGCBounds (2.93s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRestoreAsOfSystemTimeGCBounds.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

backup.TestRestoreAsOfSystemTimeGCBounds [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20988143?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20988143?buildTab=artifacts#/) on master @ [2311287354d2af647c5803b567fd01aea5c7f9ce](https://github.com/cockroachdb/cockroach/commits/2311287354d2af647c5803b567fd01aea5c7f9ce):


```
=== RUN   TestRestoreAsOfSystemTimeGCBounds
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/322912c8338911c28b7252db0743d5bf/logTestRestoreAsOfSystemTimeGCBounds3868645182
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:183: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    backup_test.go:3895: error executing query="BACKUP TABLE data.bank INTO $1 WITH revision_history" args=["nodelocal://1//tbl-after-gc"]: pq: batch timestamp 1768585712.245619704,0 must be after replica GC threshold 1768585712.395150591,0 (r81: /Tenant/10{-/Table/106/1/1})
    testutils.go:326: no Invalid Descriptors
    panic.go:615: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/322912c8338911c28b7252db0743d5bf/logTestRestoreAsOfSystemTimeGCBounds3868645182
--- FAIL: TestRestoreAsOfSystemTimeGCBounds (2.43s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRestoreAsOfSystemTimeGCBounds.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/sql/catalog/lease/BUILD.bazel
+++ b/pkg/sql/catalog/lease/BUILD.bazel
@@ -98,6 +98,7 @@ go_test(
         "//pkg/jobs/jobspb",
         "//pkg/keys",
         "//pkg/kv",
+        "//pkg/kv/kvclient/kvcoord",
         "//pkg/kv/kvpb",
         "//pkg/kv/kvserver",
         "//pkg/kv/kvserver/kvserverbase",

--- a/pkg/sql/catalog/lease/lease.go
+++ b/pkg/sql/catalog/lease/lease.go
@@ -2090,11 +2090,12 @@ func (m *Manager) AcquireByName(
 	// way: look in the database to resolve the name, then acquire a new lease.
 	var err error
 	id, err := m.resolveName(ctx, timestamp.GetTimestamp(), parentID, parentSchemaID, name)
-	if err != nil &&
-		(timestamp.GetTimestamp() == timestamp.GetBaseTimestamp() ||
-			!errors.Is(err, catalog.ErrDescriptorNotFound)) {
-		return nil, err
-	} else if errors.Is(err, catalog.ErrDescriptorNotFound) {
+	if err != nil {
+		if timestamp.GetTimestamp() == timestamp.GetBaseTimestamp() ||
+			(!errors.Is(err, catalog.ErrDescriptorNotFound) &&
+				!errors.HasType(err, (*kvpb.BatchTimestampBeforeGCError)(nil))) {
+			return nil, err
+		}
 		// The descriptor was not found at the lease timestamp, so attempt the
 		// real timestamp in use by this txn. This implies the object may have
 		// just been created.
@@ -3735,4 +3736,12 @@ func (m *Manager) GetReadTimestamp(ctx context.Context, timestamp hlc.Timestamp)
 // TestingGetBoundAccount returns the bound account used by the lease manager.
 func (m *Manager) TestingGetBoundAccount() *mon.ConcurrentBoundAccount {
 	return m.boundAccount
+}
+
+// TestingGetGCLeaseTimestamp returns a timestamp that will intentionally be GCed.
+func TestingGetGCLeaseTimestamp(readTimestamp hlc.Timestamp) ReadTimestamp {
+	return &LeaseTimestamp{
+		ReadTimestamp:  readTimestamp,
+		LeaseTimestamp: hlc.Timestamp{}.Next(),
+	}
 }

--- a/pkg/sql/catalog/lease/lease_test.go
+++ b/pkg/sql/catalog/lease/lease_test.go
@@ -26,6 +26,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
 	"github.com/cockroachdb/cockroach/pkg/keys"
 	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvserverbase"
@@ -4136,4 +4137,62 @@ func TestLeaseManagerLockedTimestampRenames(t *testing.T) {
 		})
 		require.NoError(t, grp.Wait())
 	})
+}
+
+// TestLeaseBeforeGC validates if a lease timestamp for some reason is before
+// a GC interval, we correctly fallback to the read timestamp from the txn.
+func TestLeaseBeforeGC(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	st := cluster.MakeTestingClusterSettings()
+	ctx := context.Background()
+	lease.LockedLeaseTimestamp.Override(ctx, &st.SV, true)
+	lease.WaitForInitialVersion.Override(ctx, &st.SV, false)
+
+	ts, conn, _ := serverutils.StartServer(
+		t, base.TestServerArgs{
+			Settings: st,
+		})
+	defer ts.Stopper().Stop(ctx)
+
+	runner := sqlutils.MakeSQLRunner(conn)
+	// Drop the table and GC the system tables involved.
+	runner.Exec(t, "CREATE TABLE t1(n int)")
+	var oldID descpb.ID
+	row := runner.QueryRow(t, `SELECT 't1'::REGCLASS::OID`)
+	row.Scan(&oldID)
+	runner.Exec(t, "DROP TABLE t1")
+	runner.Exec(t, "SELECT crdb_internal.unsafe_delete_descriptor($1);", oldID)
+	{
+		gcr := kvpb.GCRequest{
+			RequestHeader: kvpb.RequestHeader{
+				Key:    keys.SystemSQLCodec.TablePrefix(keys.SystemDatabaseID),
+				EndKey: keys.MaxKey,
+			},
+			Threshold: ts.Clock().Now(),
+		}
+		if _, err := kv.SendWrapped(
+			ctx, ts.SystemLayer().DistSenderI().(*kvcoord.DistSender), &gcr,
+		); err != nil {
+			t.Fatal(err)
+		}
+	}
+	// Recreate the user table.
+	runner.Exec(t, "CREATE TABLE t1(n int)")
+	row = runner.QueryRow(t, `SELECT "parentID", "parentSchemaID" FROM system.namespace WHERE name='t1'`)
+	var parentID, parentSchemaID descpb.ID
+	row.Scan(&parentID, &parentSchemaID)
+
+	now := ts.Clock().Now()
+	// Intentionally ask the user table with a timestamp that
+	// will be GCed.
+	lm := ts.LeaseManager().(*lease.Manager)
+	rs := lease.TestingGetGCLeaseTimestamp(now)
+	systemDesc, err := lm.AcquireByName(ctx, rs, parentID, parentSchemaID, "t1")
+	require.NoError(t, err)
+	defer systemDesc.Release(ctx)
+	systemDesc, err = lm.Acquire(ctx, rs, systemDesc.GetID())
+	require.NoError(t, err)
+	defer systemDesc.Release(ctx)
 }

